### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by andreacortis** [pymoca/pymoca#236] [General]
Hello, I just found this and I think an incredibly beautiful project!
Would the developers care to comment on what is the roadmap? what are the still the main drawbacks? 

Best
Andrea

**Comment by kentrutan** [comment]
I have been trying to catch up on the GitHub project history, but a lot of it is dated.

@jackvreeken can you give us some direction? Would be great to help me plan my work.

**Reply by jackvreeken** [reply]
I totally missed this discussion, so thanks for notifying me!

As far as direction or roadmap is concerned, I don't think there really is one. If the past is any indication, people tend to work on what _they_ find fun and useful. I certainly don't want to dictate what others should work on, especially for a project this small.

The list of issues should at least provide an overview of Pymoca's shortcomings, but none of them are urgent enough for me personally to work on them (at time of writing anyway). That does not mean they can't be very valuable to somebody else, and I very much encourage you to work on what would be most useful to you (and/or your direct colleagues).

Some "themes" of gaps in Pymoca are:
- Support for various Modelica language constructs (discrete events, types, see e.g. #4 and #66)
- Support for more/different backends. SymPy is a bit lacking at the moment I think, but I don't really use it personally.
- I'm mostly tied to the CasADi backend, and mostly for optimal control purposes. It would be nice to be able to also simulate models using the CasADi backend, but as far as I can tell either CasADi or Pymoca has to add some type of index-reduction for that to work (with e.g. the Sundials solvers). I looked it up once, but didn't get very far in understanding what it means and how to implement it.

As far as current smaller/bigger annoyances of me personally are concerned:
- The tests are a bit slow. That is because for many tests, the entire model folder (all "*.mo" files) is parsed time and again. Later on we started to use inline model definitions a bit more, but there is room for improvement here.
- Related to that, parsing is _slow_! This is mostly because of ANTLR, and we can't really improve on that unless we somehow let the parsing be handled by a compiled/C(++) library. Caching of files (not models) may help though.
- Having a more low-level parsing/flattener/ast in a C/Rust library will solve the performance issue, and also provide more interoperability to other languages. I don't really know how much interest there is to use Pymoca outside of the Python ecosystem though. I don't want to give up Python just for speed, as it means giving up a lot of flexibility, and easy of programming and debugging.

The wider field of domain specific languages for modelling may also be a source of inspiration. The Julia community in particular seems to be working a lot on these topics, see e.g. Modia.jl and Symbolics.jl.

**Reply by kentrutan** [reply]
@jackvreeken Thanks! Very helpful.

**Comment by kentrutan** [comment]
The README now has an updated roadmap section.
